### PR TITLE
[bazel][python] Support _virtual_imports input for py_proto_library and py_grpc_library rules

### PR DESCRIPTION
**Pull request by vam-google**
(This change is needed by the [googleapis](https://github.com/googleapis/googleapis) repository)

The `proto_library` targets are used as deps for `py_proto_library` and `py_grpc_library` rules. The `proto_library` targets can be configured using `import_prefix` and/or `strip_import_prefix` (which essentially move original location of the proto file and, as result, affects import path within proto files themselves and in the generated language-specific stubs).

The biggest question to answer when generating stubs from moved protos is where to put the result (this decision affects all downstream rules as well, because the location of a file affects its import path).

**tl;dr; This PR tries to follow same logic as the native `cc_proto_library` (created and maintained by Bazel team).**  More details below.

For example, if we have `firestore.proto` file, which is located under `google/firestore/v1beta1` Bazel package and want to move it under `google/cloud/firestore_v1beta1/proto` (this is axactly what happens in googleapis repository), it can be configured the following way:

```bzl
proto_library(
    name = "firestore_moved_proto",
    srcs = ["firestore.proto"],
    import_prefix = "google/cloud/firestore_v1beta1/proto",
    strip_import_prefix = "google/firestore/v1beta1",
)
```

The rule above will first generate virtual `.proto` files (under new location) and only after that generate a binary descriptor from them.
Specifically it will generate the following "virtual" file under `_virtual_imports` subdirectory of same package:

```
bazel-bin/google/firestore/v1beta1/_virtual_imports/firestore_moved_proto/google/cloud/firestore_v1beta1/proto/common.proto
```

When supplied to `cc_proto_library`, like the following:

```bzl
cc_proto_library(
    name = "firestore_moved_cc_proto",
    deps = ["firestore_moved_proto"],
)
```

The rule will generate .cc and .h files like the following:
```
bazel-bin/google/firestore/v1beta1/_virtual_imports/firestore_moved_proto/google/cloud/firestore_v1beta1/proto/firestore.pb.h
bazel-bin/google/firestore/v1beta1/_virtual_imports/firestore_moved_proto/google/cloud/firestore_v1beta1/proto/firestore.pb.cc
```

Notice, that each generated `.cc` and `.h` file is prefixed with `_virtual_imports/<name_of_proto_library_target>`.

The python rules try to do same thing, so for the following py_proto_library rule:

```bzl
py_proto_library(
    name = "firestore_moved_py_proto",
    deps = [":firestore_moved_proto"],
)
```

It wil generate the following file:
```
bazel-bin/google/firestore/v1beta1/_virtual_imports/firestore_moved_proto/google/cloud/firestore_v1beta1/proto/firestore_pb2.py
```

I.e in same path as cc_proto_library.

This all woks, but an annoying part is that to use the generated library in some other py_library, in needs to specify the "_virtual_imports/proto_target" path as its "includes" parameter (and I can't make it better than that).

**Another option would be to skeep** the `_virtual_imports/<name_of_proto_library_target>` the suffix for the generated python stubs, which will make the path look like the following:
```
bazel-bin/google/firestore/v1beta1/google/cloud/firestore_v1beta1/proto/firestore_pb2.py
```

That will make using generated stubs simpler and cleaner (no need to specify imports argument), but it will make it inconsistent with the other rules (like cc_proto_library) and also more susseptible to naming conflicts (if there is already something under the generated path).

@mhaidrygoog

### CODE REVIEW

**Comment by vam-google** [comment]
@gnossen PTAL (i'm happy to schedule a gvc to discuss the details)

**Comment by vam-google** [comment]
@eoogbe FYI

File: bazel/test/python_test_repo/BUILD [hunk 1]
```diff
@@ -60,3 +60,54 @@ py_test(
...
+    main = "helloworld.py",
+    srcs = ["helloworld.py"],
+    deps = [
+        ":helloworld_moved_py_pb2",
+        ":helloworld_moved_py_pb2_grpc",
+        ":duration_py_pb2",
+        ":timestamp_py_pb2",
+    ],
+    imports = [
+        "_virtual_imports/helloworld_moved_proto",
```
Comment (gnossen) [inline bazel/test/python_test_repo/BUILD]: I somewhat wonder if it's considered good practice to depend on the `_virtual_imports` name. Is this part of Bazel's API?  Or is it considered an implementation detail? Who would be able to answer this question for us? @lberki perhaps?

File: bazel/test/python_test_repo/.bazelrc [hunk 2]
```diff
@@ -0,0 +1,3 @@
+build:python3 --python_path=python3
```
Comment (gnossen) [inline bazel/test/python_test_repo/.bazelrc]: Just wanted to add that there's [a PR](https://github.com/grpc/grpc/pull/20124/files) up right now that might result in merge conflicts with this.

File: bazel/generate_cc.bzl [hunk 3]
```diff
@@ -6,7 +6,7 @@ directly.
 
 load(
     "//bazel:protobuf.bzl",
-    "get_include_protoc_args",
+    "get_include_directory",
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/generate_cc.bzl]: I think in "first version" of this change it this was required (I do not remember details), but now, when I look at it, yes, I could leave `get_include_protoc_args` untouched. 

I'm fine with putting `get_include_protoc_args` back, but the main reasons why i would like to keep it as is:
- `get_include_protoc_args` did not (and could not) specify `--proto_path` completely (only a portion of it). Notice how every time `get_include_protoc_args` was called there were subsequent explicit additions to `--proto_path` rigth after it. In other words, it is hard for me to find a clear area of responsibility of this method. This method encapsulated the `--proto_path` argument, but it was not doing it completely. Extracting the gist of this method (i.e. `get_include_directory`) in its own top-level method, allows us to specify `--proto_path` consistently everywhere (i.e. `--proto_path` argument name is specified explicitly in the rule implementation multiple times, but the values of the arguments are calculated by functions)
- In general its seems like the current pattern in the `grpc/bazel` rules implementaiton is: calculation of argument values is done in shared functions (like `get_include_directory`) but the actual name of the argument (like `--proto_path`, `--python_out`, `--cpp_out` etc) is specified esplicitly in each rule implementation. There is an exception though - the `--plugin` argument.

In other words, I'm ok with reverting this, but unless you have strong opinion about it, I would prefer to keep it as is. Please let me know what you think about it.

File: bazel/protobuf.bzl [hunk 4]
```diff
@@ -153,19 +155,75 @@ def get_proto_arguments(protos, genfiles_dir_path):
...
+    return context.genfiles_dir.path
+
+def is_in_virtual_imports(source_file, virtual_folder = _VIRTUAL_IMPORTS):
+    """Determines if source_file is virtual (is placed in _virtual_imports
+    subdirectory). The output of all proto_library targets which use
+    import_prefix  and/or strip_import_prefix arguments is placed under
+    _virtual_imports directory.
+
+    Args:
+        context: A ctx object for the rule.
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/protobuf.bzl]: Fixed

File: bazel/protobuf.bzl [hunk 5]
```diff
@@ -153,19 +155,75 @@ def get_proto_arguments(protos, genfiles_dir_path):
     arguments = []
     for proto in protos:
         massaged_path = proto.path
-        if massaged_path.startswith(genfiles_dir_path):
+        if is_in_virtual_imports(proto):
+            incl_directory = get_include_directory(proto)
+            if massaged_path.startswith(incl_directory):
+                massaged_path = massaged_path[len(incl_directory) + 1:]
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/protobuf.bzl]: I'm not sure I understand what you mean here, can you please clarify it? Line 161 and 163 do same string manipulation, but have different arguments (161 uses incl_directory but 163 uses genfiles_dir_path). Also this line is on the second level `if` indentation and 163 is on the first level.

I converted the logic in such way, that the if statements determine the length of the prefix to strip, and prefix stripping happens only once.

File: bazel/protobuf.bzl [hunk 6]
```diff
@@ -153,19 +155,75 @@ def get_proto_arguments(protos, genfiles_dir_path):
...
 def declare_out_files(protos, context, generated_file_format):
     """Declares and returns the files to be generated."""
+
+    out_file_paths = []
+    for proto in protos:
+        if not is_in_virtual_imports(proto):
+            out_file_paths.append(proto.basename)
+        else:
+            path = proto.path[proto.path.index(_VIRTUAL_IMPORTS) + 1:]
+            # TODO: uncomment if '.' path is chosen over
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/protobuf.bzl]: This is a effectively my question to you: Which one should we prefer (the uncommented path calculation, or the commented). Please check the PR description for more details:
---------------------------------
**Another option would be to skeep** the `_virtual_imports/<name_of_proto_library_target>` the suffix for the generated python stubs, which will make the path look like the following:
```
bazel-bin/google/firestore/v1beta1/google/cloud/firestore_v1beta1/proto/firestore_pb2.py
```

That will make using generated stubs simpler and cleaner (no need to specify imports argument), but it will make it inconsistent with the other rules (like cc_proto_library) and also more susseptible to naming conflicts (if there is already something under the generated path).
---------------------------------

To turn on the "other" option we would need to uncomment the TODO's. If we choose to keep the option 1 (i.e. the current state of this PR), I'll just delete the TODO's and the commented code segments they belong to.

File: bazel/protobuf.bzl [hunk 7]
```diff
@@ -153,19 +155,75 @@ def get_proto_arguments(protos, genfiles_dir_path):
...
+                out_file_path,
                 generated_file_format,
             ),
         )
-        for proto in protos
+        for out_file_path in out_file_paths
     ]
+
+def get_out_dir(protos, context):
+    """ Returns the calcualted value for --<lang>_out= protoc argument based on
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/protobuf.bzl]: Fixed

File: bazel/protobuf.bzl [hunk 8]
```diff
@@ -153,19 +155,75 @@ def get_proto_arguments(protos, genfiles_dir_path):
...
+    """
+    at_least_one_virtual = 0
+    for proto in protos:
+        if is_in_virtual_imports(proto):
+            at_least_one_virtual = True
+        elif at_least_one_virtual:
+            fail("Proto sources must be either all virtual imports or all real")
+    if at_least_one_virtual:
+        return get_include_directory(protos[0])
+        # TODO: uncomment if '.' path is chosen over
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/protobuf.bzl]: Plese check my response for the other wone TODO.

File: bazel/test/python_test_repo/BUILD [hunk 9]
```diff
@@ -60,3 +60,54 @@ py_test(
...
+# strip_import_prefix arguments
+proto_library(
+    name = "helloworld_moved_proto",
+    srcs = ["helloworld.proto"],
+    deps = [
+        "@com_google_protobuf//:duration_proto",
+        "@com_google_protobuf//:timestamp_proto",
+    ],
+    import_prefix = "google/cloud",
+    strip_import_prefix = ""
```
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/test/python_test_repo/BUILD]: `strip_import_prefix` I think `strip_import_prefix = ""` is same as not specifying it at all, but I just put it here to make it more explicit (and if somebody searches for this prefix as an example they may find the relevant test case). To put something reasonable in `strip_import_prefix` we need to have some prefix (note that this happens on top-level of the `python_test_repo` "sub-workspace" so there is no any prefix to strip). `strip_import_prefix` and `import_prefix` are about same thing, so testing one one of them should be enough. Otherwise I would need to create subpackage here with new protos and stuff... it just felt like too much of unnecessary garbage in this PR).

File: bazel/test/python_test_repo/BUILD (hunk 1 above)
Comment (vam-google) [reply bazel/test/python_test_repo/BUILD]: I think it is strictly a bad practice... Enabling the TODOs will allow us to avoid doing this, but with cost of making `py_proto_library` inconsistent with `cc_proto_library`.

Another option can be adding this iports inside `native.py_library` call under `py_proto_library`.

File: bazel/generate_cc.bzl (hunk 3 above)
[the remark this answers is not in the document]
Comment (gnossen) [reply bazel/generate_cc.bzl]: That's a fair argument. It's fine the way you have it now.

File: bazel/test/python_test_repo/.bazelrc (hunk 2 above)
Comment (vam-google) [reply bazel/test/python_test_repo/.bazelrc]: Can you please clarify your comment? This is `.bazelrc` file for the test "sub-workspace". 

This is needed, because `py_proot_library` rules break on our workstations with bazel 0.27.0+.Starting from this version Bazel started resolving python differently and instead of resolving to Python 2.7 by default it resolves to Python 3.6. This breaks python builds, because of the `enum34` dependency (imported here https://github.com/grpc/grpc/blob/master/bazel/grpc_python_deps.bzl#L25 (which is not supported after Python3.4 or something and conflicts with the built in enum in python)).

Comment (gnossen) [reply bazel/test/python_test_repo/.bazelrc]: Right. We dealt with this problem in the PR I linked above. And part of the solution is to explicitly removing the usage of the `python_path` flag. All that should be necessary now is to remove the subworkspace's `.bazelrc` and to explicitly add a `python_version` attribute to the tests.

The diff is a bit messed up in the unreversion PR I linked, but in [the original PR](https://github.com/gnossen/grpc/blob/e70788364bf38947e9d3afb820800b14e44829b9/bazel/python_rules.bzl#L183), we switched to using a custom `py2and3_test` for every test.

File: bazel/test/python_test_repo/BUILD (hunk 1 above)
Comment (gnossen) [reply bazel/test/python_test_repo/BUILD]: If `cc_proto_library` already requires users to reference `_virtual_imports` in their `#include`s, that sounds as good as an API commitment from the Bazel team, but I would like to hear back from them before we make a decision.

Comment (vam-google) [reply bazel/test/python_test_repo/BUILD]: I don't know if _virtual_imports becomes visible for people or no (I would guess no). I know that this is where cc_proto_library puts its generated files, but users of the lib probably import it from different location (cc_proto_library is a native rule so it can do weird things, also C++ is compilable, so sources are just sources, and not used directly, compiled binary is used instead, which is different for python (where source and library are same thing)).

Comment (lberki) [reply bazel/test/python_test_repo/BUILD]: Adding our Python expert, @brandjon for a more authoritative answer.

I'd much rather no one encodes `_virtual_imports` in BUILD files. That won't change (for a while), but it's considered an implementation detail and it's at Bazel's discretion whether it creates such a directory for any given `proto_library` rule.

The "official" way to do `LANG_proto_library` is to make them not macros, but rules (i.e. not `def bf_proto_library()` but `bf_proto_library = rule(...)`.) Then they can create language-specific providers (`PyInfo` in this case) and are able to do things like adding a new entry to the import path and are able to figure out he proto source root (by calling `dep[ProtoInfo].proto_source_root`). That way, dependent Python rules don't need to know how exactly they work.

Comment (vam-google) [reply bazel/test/python_test_repo/BUILD]: @lberki 

tl;dr; Is it possible to create PyInfo and or ProtoInfo provider instance from a Starlark rule? Last time I tried to create ProtoInfo from starlark I could not figure out how to do it and concluded that it could be created only from navite rules (but it was long time ago and I may be wrong here).

`py_proto_library` here consists of essentially two calls only: 1) generate python stubs, 2) call native py_library rule to make it a "library" (compile, if python was static languge). 

So it is as simple wrapper macro as it can be, and probably it is done this way because it allows to output a legit `py_library` without need to re-implement `py_library` here. 

`py_library` is a native rule with opaque implementation, so it is hard to figure out how to make py_proto_library in a form of a rule such that it can be consumed as a dependency (by `py_library`, `py_binary` and `py_test`).

File: bazel/test/python_test_repo/.bazelrc (hunk 2 above)
Comment (vam-google) [reply bazel/test/python_test_repo/.bazelrc]: Switched to py2and3test and removed `.bazelrc` file from the build

File: bazel/test/python_test_repo/BUILD (hunk 1 above)
Comment (vam-google) [reply bazel/test/python_test_repo/BUILD]: I have update the generation rule to return `PyInfo` provider as well with imports property set to point to `_virtual_imports`, and then pass that as deps to the subsequent py_library rule. This effectively allowed to hide `_virtual_imports` from the surface.

File: bazel/protobuf.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (vam-google) [reply bazel/protobuf.bzl]: This is not needed anymore and was removed.

**Comment by gnossen** [comment]
`rubygems.org` appears to be down, causing just about all the tests to fail... Trying again.
